grading/execute.cpp

```diff
@@ -39,150 +39,44 @@ extern const int CPU_TO_WALLCLOCK_TIME_BUFFER;  // defined in default_config.h
-    { "ocamlyacc",               "/usr/bin/ocamlyacc" },
-    { "z3",                      SUBMITTY_INSTALL_DIRECTORY+"/tools/z3" },
-
-    // for Cmake & Make
-    { "cmake",                   "/usr/bin/cmake" },
-    { "make",                    "/usr/bin/make" },
-
-    // for Network Programming
-    { "timeout",                 "/usr/bin/timeout" },
-    { "mpicc.openmpi",           "/usr/bin/mpicc.openmpi" },
-    { "mpirun.openmpi",          "/usr/bin/mpirun.openmpi" },
-    { "mpirun",                  "/usr/local/mpich-3.2/bin/mpirun"},
-    { "mpicc",                   "/usr/local/mpich-3.2/bin/mpicc"},
-    { "expect",                  "/usr/bin/expect" },
-    { "sleep",                   "/bin/sleep" },
-
-    // for Distributed Systems
-    { "script",                  "/usr/bin/script" },
-
-    // for LLVM / Compiler class
-    { "lex",                     "/usr/bin/lex" },
-    { "flex",                    "/usr/bin/flex" },
-    { "yacc",                    "/usr/bin/yacc" },
-    { "bison",                   "/usr/bin/bison" },
-
-    // for graphics/window interaction
-    { "scrot",                   "/usr/bin/scrot"}, //screenshot utility
-    { "xdotool",                 "/usr/bin/xdotool"}, //keyboard/mouse input
-    { "wmctrl",                  "/usr/bin/wmctrl"}, //bring window into focus
-    { "xwininfo",                "/usr/bin/xwininfo"}, // get info about window
-
-    // for Debugging
-    { "strace",                  "/usr/bin/strace" },
-
-    //Matlab
-    { "matlab",                  "/usr/local/bin/matlab" }
-
-  };
-
-  if(running_in_docker){
-    allowed_system_programs.insert({"bash", "/bin/bash"});
-    allowed_system_programs.insert({"php",  "/usr/bin/php"});
-  }
-  // find full path name
-  std::map<std::string,std::string>::const_iterator itr = allowed_system_programs.find(program);
-  if (itr != allowed_system_programs.end()) {
-    full_path_executable = itr->second;
-    return true;
-  }
-
-  // did they already use the full path name?
-  for (itr = allowed_system_programs.begin(); itr != allowed_system_programs.end(); itr++) {
-    if (itr->second == program) {
-      full_path_executable = program;
-      return true;
+bool system_program(const std::string &program, std::string &full_path_executable, const bool running_in_docker)
+{
+    nlohmann::json allowed_system_programs;
+    std::stringstream default_sstr(GLOBAL_default_allowed_commands_string);
+    default_sstr >> allowed_system_programs;
+    std::stringstream custom_sstr(GLOBAL_custom_allowed_commands_string);
+    if (!custom_sstr.str().empty()) {
+        nlohmann::json allowed_system_programs_custom;
+        custom_sstr >> allowed_system_programs_custom;
+        allowed_system_programs.update(allowed_system_programs_custom);
     }
-  }
-
-  // not an allowed system program
-  return false;
+    if(running_in_docker){
+        allowed_system_programs["bash"] = "/bin/bash";
```

Comment:
maybe not in this PR (future PR?)
But it would be good to put this list of commands into the json file(s) as well
Either a separate json with commands for docker containers only, or make the json structure more complex with a boolean per flag about whether it available only inside docker container autograding.  
A more complex json structure would have the benefit of easy extensibility for future features.

Reply:
Items now have a structure such as this.

```
"php":                     { "path": "/usr/bin/php", "onlyDocker": true }
"matlab":                  { "path": "/usr/local/bin/matlab" },
```